drivers/sensor/nuvoton_adc_cmp_npcx/adc_cmp_npcx.c

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/sensor.h>
+#include <drivers/adc/adc_npcx_threshold.h>
+#include <drivers/sensor/adc_cmp_npcx.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(adc_cmp_npcx, CONFIG_SENSOR_LOG_LEVEL);
+
+struct adc_cmp_npcx_data {
+	/* Work queue to be notified when threshold assertion happens */
+	struct k_work work;
+	/* Sensor trigger hanlder to notify user of assetion */
+	sensor_trigger_handler_t handler;
+	/* ADC NPCX driver reference */
+	const struct device *dev;
+};
+
+struct adc_cmp_npcx_config {
+	/*
+	 * Pointer of ADC device that will be performing measurement, this
+	 * must be provied by device tree.
+	 */
+	const struct device *adc;
+	/*
+	 * ADC channel that will be used to measure signal, this must be
+	 * provided by device tree.
+	 */
+	uint8_t chnsel;
+	/* Threshold selection number assigned during initialization */
+	uint8_t th_sel;
+	/* Threshold assert value in millivolts */
+	uint32_t thr_mv;
+	/*
+	 * Condition to be met between measured signal and threshold assert
+	 * value that will trigger event
+	 */
+	enum adc_cmp_npcx_comparison comparison;
+};
+
+#define DT_DRV_COMPAT nuvoton_adc_cmp
+
+#define ADC_CMP_NPCX_UNDEFINED		(-1)
+static void adc_cmp_npcx_trigger_work_handler(struct k_work *item)
+{
+	struct adc_cmp_npcx_data *data =
+			CONTAINER_OF(item, struct adc_cmp_npcx_data, work);
+	struct sensor_trigger trigger = {
+		.type = SENSOR_TRIG_THRESHOLD,
+		.chan = SENSOR_CHAN_VOLTAGE
+	};
+
+	if (data->handler) {
+		data->handler(data->dev, &trigger);
+	}
```

Comment:
You need to store and pass the pointer to the `struct sensor_trigger` provided when registering the trigger in order for client code to use `CONTAINER_OF()` in the trigger callback function.

Reply:
Using the caller's provided `struct sensor_trigger` in a follow-on PR seems reasonable.  `adc_cmp_npcx_trigger_set()` already guards that the trigger type is `SENSOR_TRIG_THRESHOLD` and the channel is `SENSOR_CHAN_VOLTAGE`.

Reply:
> That certainly is a useful feature, however, this PR has been open for quite some time and that change will require client code modifications.

Even if we accept the current form, someone could submit a PR the day after this, fixing the bug - which will then require client code modifications.

> Could we have that feature implemented in a follow up PR and move forward with current one?

We can do that. Please open a follow-up PR for that as soon as this PR is merged.

Reply:
Thanks, I will do that once this is merged.